tests/virt_autotest/sriov_network_card_pci_passthrough.pm

```diff
@@ -0,0 +1,368 @@
+    prepare_host();
+
+    #get the SR-IOV device BDF and interface
+    my @host_pfs;
+    @host_pfs = find_sriov_ethernet_devices();
+    if (@host_pfs == ()) {
+        $self->{test_results}->{host}->{"Error: there is no SR-IOV ethernet devices in the host!"}->{status} = 'FAILED';
+        return 1;
+    }
+    record_info("Find SR-IOV devices", "@host_pfs");
+
+    #get/set nessisary variables for test
+    my $gateway = script_output "ip r s | grep 'default via' | cut -d ' ' -f3";
+    set_var("SRIOV_NETWORK_CARD_PCI_PASSSHTROUGH", 1);    #to differenciate virtual network tests
+
+    # enable 8 vfs for the SR-IOV device on host
+    my @host_vfs = enable_vf(@host_pfs);
+    record_info("VFs enabled", "@host_vfs");
+
+    foreach my $guest (keys %xen::guests) {
+
+        prepare_guest($guest);
+        upload_network_device_status_logs($guest, "01-initial");
+
+        #passthrough 2 vf ethernet devices from host
+        my @vfs               = ();
+        my $passthru_vf_count = 2;    #the number of vfs to be passed through to guests
+        for (my $i = 0; $i < $passthru_vf_count; $i++) {
+
+            my %vf;
+
+            #detach the vf from host
+            $vf{host_bdf} = $host_vfs[int(rand($#host_vfs + 1))];
+            for (my $j = 0; $j < $i; $j++) {
+                if ($vf{host_bdf} eq $vfs[$j]->{host_bdf}) {
+                    $vf{host_bdf} = $host_vfs[int(rand($#host_vfs + 1))];
+                    $j = 0;
+                }
+            }
+            $vf{host_id} = detach_vf_from_host($vf{host_bdf});
+
+            #add the vf to the list of passthrough devices
+            push @vfs, \%vf;
+
+            #plug the vf in guest
+            plugin_device($guest, $vfs[$i]);
+
+            #upload test specific logs
+            upload_network_device_status_logs($guest, "02-after_hotplug_$vfs[$i]->{host_id}");
+
+            #check the networking of the plugged interface
+            #use br123 as ssh connection
+            test_network_interface($guest, gate => $gateway, mac => $vfs[$i]->{vm_mac}, net => 'br123');
+
+        }
+
+        #reboot the guest
+        record_info("VM reboot", "$guest");
+        script_run "ssh root\@$guest 'reboot'";    #don't use assert_script_run, or may fail on xen guests
+        upload_network_device_status_logs($guest, "03-after_guest_reboot");
+        script_retry("nmap $guest -PN -p ssh | grep open", delay => 10, retry => 18, die => 1);
+
+        #check again the networking inside vm
+        for (my $i = 0; $i < $passthru_vf_count; $i++) {
+            test_network_interface($guest, gate => $gateway, mac => $vfs[$i]->{vm_mac}, net => 'br123');
+        }
+
+        #unplug the first vf from vm
+        unplug_vf_from_vm($guest, $vfs[0]);
+        record_info("VF unpluged from vm", "$vfs[0]->{host_id} \nGuest: $guest \nbdf='$vfs[0]->{vm_bdf}'   mac='$vfs[0]->{vm_mac}'   nic='$vfs[0]->{vm_nic}'");
+        assert_script_run("virsh nodedev-reattach $vfs[0]->{host_id}", 60);
```

Comment:
I would add a simple test to check, if the NIC has been detached successfully here. Checking `ip link` might be already sufficient. Otherwise we rely only on the return value of `virsh detach-device` (in `sub unplug_vf_from_vm`) and do not check, if it has been actually unplugged.

Reply:
Yes, you are right. I should do some check in sub unplug_vf_from_vm.